### DISCUSSION TITLE
Configurable trailing slash preference in URLs

### DISCUSSION THREAD

**Discussion by hrumhurum** [retypeapp/retype#219] [Q&A] [answered]
We spotted that Retype has a strong preference to put a trailing slash at the end of the URLs, like this
`https://retype.com/components/file-download/`.

However our standard is to use slash-less URLs unless they point to a directory, like so `https://retype.com/components/file-download`. 

Is there a way to configure that preference?

**Comment by DannyBen** [comment]
I believe is not Retype's fault, this is just the best (arguably?) way to generate static sites.

Retype generates a static HTML site, with everything (files or folders) tucked inside a `folder/index.html` pattern.

So for example, a file named `installation.md` is generated as `installation/index.html` and not `installation.html`.

Generating it any other way, would mean that the URLs will have to include the `.html` extension, which is much less desirable.

**Reply by AndreyChechel** [reply]
I think, possibility to refer a page without a trailing slash depends on a web server used to host a Retype website. Say, for `retype watch` we use Kestrel server and it does resolve route to a `/index.html` page correctly, so inclusion of a trailing slash becomes more like a preference, not a must.

Even more - early versions of Retype didn't add a trailing slash to resolved URLs. However, most web servers we'd seen generated 301 redirects to land users on a canonical URL suffixed with a slash, so we revised the logic.

Provided it depends on a particular web server settings, I see practical use of having a configuration option to control trailing slash emission. Current behavior will still be the default one.

**Reply by DannyBen** [reply]
I think `retype watch` is of no consequence to this discussion, as it is a development-only component.

Since retype is a static site generator, it must adhere to static HTML standards, and its output is expected to work in any static web server. Which means one of these options:

1. URLs end with a slash, and point to an index.html in that folder. Note that in this case, all webservers I know will redirect `domain/folder` to `domain/folder/` so slash-less URLs also work, but will always show the slash in the end URL.
2. URLs end with a file extension, so the webserver knows that `domain/index.html` is of `text/html` mime type. This of course is widely regarded as bad practice for HTML pages these days.
3. Use javascript to hijack clicks, and perform whatever they want in terms of loading new content and injecting a new URL.

Option 3 is terrible, but is being used by all sorts of javascript frameworks.

I would be curious to see if you have a way to still maintain a standard static site (without javascript acrobatics) and have slash-less URLs.

**Reply by AndreyChechel** [reply]
Well, Kestrel is still a very good example. It has [out-of-the box configuration](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.staticfiles.infrastructure.sharedoptionsbase.redirecttoappendtrailingslash?view=aspnetcore-6.0#Microsoft_AspNetCore_StaticFiles_Infrastructure_SharedOptionsBase_RedirectToAppendTrailingSlash) for forcing a trailing slash in requests to static resources/pages. Kestrel is widely used in many production environments for ASP.NET Core app serving.

**Reply by AndreyChechel** [reply]
For instance, [here](https://stackoverflow.com/a/31189870/7889645) people discuss how to configure Nginx to serve `/index.html` when there is no trailing slash included. So it certainly depends on routing capabilities of a web server and what configuration it has. Anyways I totally agree, that **by default** a trailing slash is expected by most of web servers.

**Comment by AndreyChechel** [comment, marked as answer]
Hi @hrumhurum 

Currently there is no configuration for that, but that should be possible for us to implement, just need to ensure the rule can be applied consistently across all website pages (resolvable user links, generatable links like tags / categories / next, prev buttons etc.).

Thanks for the feature suggestion!

**Reply by DannyBen** [reply]
How can you generate URL that does not have a slash and does not require an extension suffix in a static website? Static webservers will not know the mime type of a `domain.com/file`.

**Reply by AndreyChechel** [reply]
Please see my reply on your previous comment. I can't say for all web servers, but they will likely probe several options in order to resolve a URL, e.g.:
```
domain.com/file
domain.com/file.html
domain.com/file/index.html
```

depending on the configuration, they can try more default extensions (e.g. like `htm`, `aspx` ), and more default page names, like `index`, `default` etc. I can see that for some websites this behavior can be undesired resulting in ambiguous references. But it should be ok for Retype-generated websites, since all pages are always emitted using `name/index.html` structure pattern.

**Reply by DannyBen** [reply]
You should not assume that all users of retype control the web server, or can choose it. We are hosting our sites in different places, sometimes with no control.

An example of a failure to serve an extension-less file, is GitHub Pages:

In [this docs folder](https://github.com/DannyBen/experiments/tree/master/docs) that contains an HTML file named [hello](https://github.com/DannyBen/experiments/blob/master/docs/hello), when you access the file through GitHub pages domain, you do not get a page, but rather a downloaded file named "hello", as can be observed here: <http://dannyben.github.io/experiments/hello>

**Reply by geoffreymcgill** [reply]
> An example of a failure to serve an extension-less file, is GitHub Pages:

Retype does not create extension-less files.

Except of course the `CNAME` file, but that file is a GitHub Pages configuration file and is not served.

**Reply by DannyBen** [reply]
> Retype does not create extension-less files.

I know. And it shouldn't in my view. But the entire discussion above presented examples of how extension-less files can be served by webservers, and might be considered an option for future retype version to solve the subject of this discussion. This is why I mentioned it as an example.

**Reply by AndreyChechel** [reply]
Thanks for making a live sample with Github pages. My assumption that `file.html` is probed **by default** for `/file` requests is likely wrong. Although, my main point was that it's doable and people may configure their web servers like that. So it's a legit question for Retype to support such configurations or not.

Just to make it clear, I'm not proposing changing the default behavior implying a trailing slash always emitted. It's more like about adding another configuration option in `retype.yml` allowing to change the default behavior.
